### DISCUSSION TITLE
emacs configuration?

### DISCUSSION THREAD

**Discussion by anarcat** [Automattic/harper#150] [Q&A] [answered]
First, this looks like a fantastic project, really impressive!

Has anyone managed to make this work in Emacs?

I have first tried to setup something in lsp-mode but couldn't figure out how to add a custom language server, so i switched to eglot (which is something i've been meaning to do for a while anyway), with the following configuration:

```elisp
(use-package eglot
  :defer
  :hook
  (python-mode . eglot-ensure)
  (markdown-mode . eglot-ensure)
  :bind (:map
         eglot-mode-map
         ("C-c e r" . eglot-rename)
         ("C-c e a" . eglot-code-actions)
         ("C-c e o" . eglot-code-action-organize-imports)
         ("C-c e d" . eldoc)
         ("C-c e f" . eglot-format)
         ("C-c e =" . eglot-format))
  :config
  (add-to-list 'eglot-server-programs
               '(markdown-mode . ("harper-ls")))
  :custom
  (eglot-autoshutdown t) ;; default is to leave servers runing when last buffer exits
  ;; (eglot-events-buffer-size 0)
  ;; (eglot-stay-out-of '(yasnippet))
  (eglot-extend-to-xref nil)) ;; cover files found through xref (M-.)
```

i *think* it works in python mode: when i switch to a python buffer for the first time, it takes a while to load (a annoying but good sign) and eventually pops a buffer with `[eglot:$dirname]` in the modeline. xref works, at least.

but in markdown mode, the server doesn't seem to start at all.  this is the diagnostic buffer in emacs:

[clipboard.txt](https://github.com/user-attachments/files/16923649/clipboard.txt)

and in `*Messages*`:

```
[jsonrpc] Server exited with status 9
Error running timer: (error "[eglot] Timed out after 30 seconds")
```

I installed harper through the `nix-bin` package manager installed on a Debian bookworm system, which *could* cause some problems, I admit, but i'm wondering if anyone else made this work here...

**Comment by sbromberger** [comment, marked as answer]
I don't know anything about emacs, but it doesn't look like you're calling `harper-ls` with the `--stdio` flag. I got timeouts when I didn't do that in my editor (helix).

**Reply by anarcat** [reply]
that was it, thanks!!! the updated, working, configuration is:

```elisp
(use-package eglot
  :defer
  :hook
  (python-mode . eglot-ensure)
  (markdown-mode . eglot-ensure)
  :bind (:map
         eglot-mode-map
         ("C-c e r" . eglot-rename)
         ("C-c e a" . eglot-code-actions)
         ("C-c e o" . eglot-code-action-organize-imports)
         ("C-c e d" . eldoc)
         ("C-c e f" . eglot-format)
         ("C-c e =" . eglot-format))
  :config
  (add-to-list 'eglot-server-programs
               '(markdown-mode . ("harper-ls" "--stdio")))
  :custom
  (eglot-autoshutdown t) ;; default is to leave servers runing when last buffer exits
  (eglot-extend-to-xref nil)) ;; cover files found through xref (M-.)
```

**Reply by elijah-potter** [reply]
This is fantastic. I'm considering setting up a Wiki page for miscellaneous configurations that people can copy/paste from. Would it be alright if I included yours?

**Reply by anarcat** [reply]
of course! i would suggest reducing it to a more bare minimum, however, something like this:

```elisp
(use-package eglot
  :hook
  (text-mode . eglot-ensure)
  :config
  (add-to-list 'eglot-server-programs
               '(text-mode . ("harper-ls" "--stdio"))))
```

everything else is kind of specific to my setup (e.g. python, keybindings, the `xref` bit), optimizations (`:defer`) or optimizations (autoshutdown)... it's not even my full config right now, which is currently:

```elisp
(use-package eglot
  :defer
  :hook
  (python-mode . eglot-ensure)
  (text-mode . eglot-ensure)
  :bind (:map
         eglot-mode-map
         ("C-c e r" . eglot-rename)
         ("C-c e a" . eglot-code-actions)
         ("C-c e o" . eglot-code-action-organize-imports)
         ("C-c e d" . eldoc)
         ("C-c e f" . eglot-format)
         ("C-c e =" . eglot-format))
  :config
  (add-to-list 'eglot-server-programs
               '(text-mode . ("harper-ls" "--stdio")))
  :custom
  (eglot-autoshutdown t) ;; default is to leave servers runing when last buffer exits
  (eglot-events-buffer-size 0) ;; disable events recording to speed
                               ;; things up, comment out to
                               ;; troubleshoot and look for the EGLOT
                               ;; buffer (eglot-stay-out-of
                               ;; '(yasnippet))
  (eglot-extend-to-xref nil)) ;; cover files found through xref (M-.)
```

that will *eventually* make it to https://gitlab.com/anarcat/emacs-d/-/blob/main/init.el once i find time to commit, if you want to cross-reference to a canonical copy.

oh, and also note that i use the `use-package` package (if that's readable at all), which is not necessarily the way everyone configures such thing. a more canonical way is documented in the [eglot docs](https://joaotavora.github.io/eglot/)  which would probably be something like:

```elisp
(with-eval-after-load 'eglot
  (add-hook 'text-mode #'eglot-ensure)
  (add-to-list 'eglot-server-programs
               '(text-mode . ("harper-ls" "--stdio"))))
```

... but i haven't tested that blob so i'm not sure it works... it should be roughly equivalent though.

**Reply by anarcat** [reply]
```lisp
  (eglot-autoshutdown t) ;; default is to leave servers runing when last buffer exits
```

for what it's worth, i'm actually having trouble with this right now. when i kill a buffer, eglot tries to "politely terminate" the LSP and fails with this:

```
[eglot] Asking EGLOT (anarcat/(notmuch-message-mode text-mode)) politely to terminate
[jsonrpc] Server exited with status 9
jsonrpc-error: "request id=2 failed:", (jsonrpc-error-code . -32602), (jsonrpc-error-message . "Unexpected params: null"), (jsonrpc-error-data)
```

this is, fwiw, what the shutdown procedure does, in case you're ready for more lisp:

```lisp
(defun eglot-shutdown (server &optional _interactive timeout preserve-buffers)
  "Politely ask SERVER to quit.
Interactively, read SERVER from the minibuffer unless there is
only one and it's managing the current buffer.

Forcefully quit it if it doesn't respond within TIMEOUT seconds.
TIMEOUT defaults to 1.5 seconds.  Don't leave this function with
the server still running.

If PRESERVE-BUFFERS is non-nil (interactively, when called with a
prefix argument), do not kill events and output buffers of
SERVER."
  (interactive (list (eglot--read-server "Shutdown which server"
                                         (eglot-current-server))
                     t nil current-prefix-arg))
  (eglot--message "Asking %s politely to terminate" (jsonrpc-name server))
  (unwind-protect
      (progn
        (setf (eglot--shutdown-requested server) t)
        (jsonrpc-request server :shutdown nil :timeout (or timeout 1.5))
        (jsonrpc-notify server :exit nil))
    ;; Now ask jsonrpc.el to shut down the server.
    (jsonrpc-shutdown server (not preserve-buffers))
    (unless preserve-buffers (kill-buffer (jsonrpc-events-buffer server)))))
```

the error buffer in emacs has this to say:

```
[client-request] (id:2) Mon Sep  9 14:18:23 2024:
(:jsonrpc "2.0" :id 2 :method "shutdown" :params nil)
[stderr] [2m2024-09-09T18:18:23.876113Z[0m [32m INFO[0m [2mtower_lsp::service::layers[0m[2m:[0m shutdown request received, shutting down
[server-reply] (id:2) ERROR Mon Sep  9 14:18:23 2024:
(:jsonrpc "2.0" :error
          (:code -32602 :message "Unexpected params: null")
          :id 2)
[internal] Mon Sep  9 14:18:23 2024:
(:message "Connection state changed" :change "killed\n")

----------b---y---e---b---y---e----------
[stderr] 
[stderr] 
[stderr] nil
[stderr] nil
[stderr] Process EGLOT (wiki/(markdown-mode text-mode)) stderr finished
```

which is pretty incomprehensible to me, but perhaps it means something to you? `*Messages*` also have this to say:

```

for now i've disabled auto-shutdown...

**Reply by yitang** [reply]
@anarcat  Thanks for providing the minimal config, it works well. 

To add to that, the `eldoc` command shows the reasons (e.g. there are two spaces), and  `eglot-code-actions` to select the suggestions (e.g. replace two spaces with one).

**Comment by sbromberger** [comment]
@elijah-potter - here's the Helix configuration. It really couldn't be more straightforward:
```
[language-server.harper-ls]
command = "harper-ls"
args = ["--stdio"]
```

**Comment by Jesse-Millwood** [comment]
Does this work with org mode? I tried the text-mode setup in the docs but I didn't see any corrections come up like when I add it to the c-mode list of eglot servers.

**Reply by yitang** [reply]
Last time I checked it works for markdown mode not org mode

**Reply by elijah-potter** [reply]
Yeah, in order for org-mode to work we need to add a parser for it inside Harper. We're open to contributions if you're familiar with the domain.

**Reply by krisbalintona** [reply]
For future readers, org-mode support has been merged: https://github.com/Automattic/harper/pull/1369

**Comment by jtpavlock** [comment]
For those like me who use lsp-mode instead of eglot, here's a related lsp-mode issue
[lsp/lsp-mode/issues/4747](https://github.com/emacs-lsp/lsp-mode/issues/4747)

**Comment by DarkWingMcQuack** [comment]
since eglot cannot run multiple lsps per buffer, i think the best solution would be to have a flymake backend for harper.

**Comment by upbqdn** [comment]
Eglot does support multiple LSP servers in a single buffer now: https://joaotavora.github.io/eglot/#Multi_002dserver-support.
